continue the discussion from #2

dotnetCarpenter opened the pull request:
Live example here: https://it-kollektivet.github.io/world/

## API

```js
.panZoom({ zoomFactor:Number }) // setup svg.panzoom.js for an element and optional define a zoomFactor (default is 0.03)
.zoom(zoomAmount, point, animate) // zoom in/out, where: zoomAmount below 1 for zooming in and above 1 to zoom out
                                  //                     point is the coordinate, in SVG user-space, to zoom from
                                  //            OPTIONAL animate is { duration, easing }
.zoomLevel() // return current zoom ratio
.zoomToOne(point, animate) // an easy way to get back to the original zoom level
```

point can be an [SVG.point](http://svgjs.com/manipulating/#point) and animate is documented here: http://svgjs.com/animating/#easing

~~You can also define a global animation via: `.panZoom({ zoomAnimate: {duration, easing} }) ` but you probably won't want that unless it's a custom function since it affects both mouse wheel zoom and programmable zoom, which can behave very different. So far I have left it as a choice for the consumer.~~

## Example

```js
var world = SVG("worldmap")
var zoomFactorOut = displayZoomLevel(document.querySelector(".info-box__zoom-factor"))

world.panZoom({ zoomFactor: 0.2 })

world.on("zoom", function() {
      zoomFactorOut(this.zoomLevel())
})

SVG.on(zoomInButton, "click", function() {
	var p = getCenterPoint(world)

	world.zoom(0.5, p, { duration: 250, easing: "<>" })
})

SVG.on(zoomOutButton, "click", function() {
	var p = getCenterPoint(world)

	world.zoom(1.5, p, { duration: 200, easing: ">" })
})

SVG.on(zoomDefaultButton, "click", function() {
      var p = getCenterPoint(world)

      world.zoomToOne(p, { duration: 400, easing: "<>" })
})

function displayZoomLevel(out) {
	return function(level) {
		out.textContent = level.toFixed(2)
	}
}

function getCenterPoint(element) {
	var size = element.node.getBoundingClientRect()
	return element.point(size.width / 2, size.height/ 2)
}
```

Fuzzyma commented on svg.panzoom.js:
```diff
@@ -129,16 +140,18 @@ SVG.extend(SVG.Doc, SVG.Nested, {
     this.on('mousedown', panStart, this, {passive:false})
     this.on('touchstart', panStart, this, {passive:false})
 
+    this.zoom = zoom
```
not a fan of this one. This should be added as a real method with extend and not monkey patching it

Fuzzyma replied:
So it would be: 
```js
SVG.extend(... {
   panZoom: function(){},
   zoom: function() {}
}
```

dotnetCarpenter replied:
I can do that but then we loose the global `zoomAnimate` option which is not a terrible idea. Since it's only useful in very edge cases, where you want to save a few lines of code to have a default custom animation function for both programmable and mouse-wheel zoom. In most cases, it's probably harmful and could possibly lead to confusion for the developer.

dotnetCarpenter commented:
I added `zoomLevel()` to get the current zoom factor, and your were right, it's more intuitive to get the ration of size to viewbox.

> Anyway I am not sure how the zoom level will help you much because it is not the ratio of size to viewbox but the ratio of old viewbox to new viewbox and that's very unintuitive

More importantly I also added `zoomToOne(point, animation)` which is a quick 'n easy way to get back to the original zoom level.

As I commented before, the global animation setting had to go (well, it didn't really had to) and I rather like that, since it's _foot gun_ for new users.

dotnetCarpenter commented:
I've updated the live example: https://it-kollektivet.github.io/world/

dotnetCarpenter commented:
In `zoomLevel` I use `.bbox()` but it would probably be more safe to use `.rbox()`. But just looking at our example in our documentation, it appears to be off. Is it just me or do we have a bug?

http://svgjs.com/geometry/#svg-rbox

Fuzzyma commented:
I would like to propose the following:

- `zoom(level, point)` - when called without an argument it returns the zoom (-level)
- `animate().zoom(level, point)`
- `zoom(level)` should use the center of the canvas as focus point as default
- panZoom caculated a zoom lvl based on its factor and uses `zoom` to do the zoom - that could be a bit of overhead, though. Or it zooms just directly in lvls... dunno

PS: You changed indentation. Never do this in a PR. Its a nightmare for reviewing. I cant see what was changed by just looking at the final changes. That really bothers me. Codestyle changes should always be another PR

PPS: To the rbox issue in the docs: Guess no element was passed to rbox. On default it gives back the clientRect in screenCoords. In the docs example we want the box in doc-coords. So it should be `rect.rbox(rect.doc())`. However: I am not a fan off using bbox here at all because what we want is simply the values of `size` in px. Bbox on svg elements always returns the tightest all enclosing bbox and not the bbox of the dimensions of the svg

dotnetCarpenter commented:
Sorry about the indentation, it was a mistake.

I didn't know that `.rbox()` could take an argument. `.doc()` just returns the parent element... Hmm, I think we are missing an explanation in the docs, but I'm too ignorant about how this works to write it.
So instead of using `.bbox()`, ~I should use `this.node.clientWidth`~? `this.node.clientWidth` is always 0 in FF. What do you propose to use here?

I'll play around with your other suggestions. Seems reasonable that animating should be done in the same manner as with all other elements in SVG.js

Fuzzyma commented on svg.panzoom.js:
```diff
@@ -1,30 +1,38 @@
+var getCenterPoint = function getCenterPoint(element) {
+  var size = element.node.getBoundingClientRect()
+  return element.point(size.width / 2,
+                       size.height/ 2)
```
you need to add x and y here to get the center

Fuzzyma commented on svg.panzoom.js:
```diff
@@ -1,30 +1,38 @@
-      var b = new SVG.Box(this.viewbox()).transform(new SVG.Matrix().scale(zoomAmount, p.x, p.y))
 
-      this.viewbox(b)
+      this.zoom(zoomAmount, p)
```
you cant zoom with a zooom amount albeit it expects a level and not a factor

Fuzzyma commented on svg.panzoom.js:
```diff
@@ -129,16 +143,50 @@ SVG.extend(SVG.Doc, SVG.Nested, {
+
+  zoom: function(zoomlevel, point) {
+    var isAnimating = this instanceof SVG.FX
+    // get target in case of the fx module, otherwise reference this
```
animation should be done seperatly

Fuzzyma commented on svg.panzoom.js:
```diff
@@ -129,16 +143,50 @@ SVG.extend(SVG.Doc, SVG.Nested, {
+    // get target in case of the fx module, otherwise reference this
+    var target = isAnimating ? this.target() : this
+
+    if(arguments.length === 0) return target.bbox().width / target.viewbox().width
```
Albeit svg nodes have a dimension, the bbox is always the closest box around all elements in it. Therefore it is not useful here.
